### ISSUE TITLE
Adding mem_limit to docker run causes VM to reboot

### ISSUE THREAD

**Issue by ibrokethecloud** [docker/docker#30015]
**Description**
Running a docker container which tries to unzip a large file on rhel 7.2 GUEST running on vmware ESXi 6.0 causes the guest to reboot.

Problem only happens on docker v1.12.2 and v1.12.5 which we have tested.

We were earlier running docker v1.10.3 on the same OS image and hypervisor and we never had the same issue.

**Steps to reproduce the issue:**
1. Please use the following dockerfile to create a image:
```
FROM ubuntu:latest
RUN dd if=/dev/urandom of=/root/random.img count=1024 bs=1M && \
                gzip /root/random.img
ENTRYPOINT gzip -d /root/random.img.gz && echo "done unzipping"
```
build the image say test:1

```
docker build -t test:1 . 
```

2. Run this container as follows:
```
docker run -it --rm -m 1000m test:1
```

**Describe the results you received:**
The VM reboots with the following messages:
```
[    0.820865] pci 0000:00:17.3: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820869] pci 0000:00:17.3: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820873] pci 0000:00:17.4: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820878] pci 0000:00:17.4: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820882] pci 0000:00:17.5: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820886] pci 0000:00:17.5: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820890] pci 0000:00:17.6: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820895] pci 0000:00:17.6: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820899] pci 0000:00:17.7: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820903] pci 0000:00:17.7: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820907] pci 0000:00:18.2: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820912] pci 0000:00:18.2: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820916] pci 0000:00:18.3: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820920] pci 0000:00:18.3: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820925] pci 0000:00:18.4: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820929] pci 0000:00:18.4: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820933] pci 0000:00:18.5: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820937] pci 0000:00:18.5: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820942] pci 0000:00:18.6: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820946] pci 0000:00:18.6: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820950] pci 0000:00:18.7: res[13]=[io  0x1000-0x0fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820954] pci 0000:00:18.7: res[13]=[io  0x1000-0x1fff] res_to_dev_res add_size 1000 min_align 1000
[    0.820971] pci 0000:00:15.0: BAR 15: assigned [mem 0xc0000000-0xc01fffff 64bit pref]
[    0.821395] pci 0000:00:0f.0: BAR 6: assigned [mem 0xc0200000-0xc0207fff pref]
[    0.822389] pci 0000:00:10.0: BAR 6: assigned [mem 0xc0208000-0xc020bfff pref]
[    0.823390] pci 0000:00:15.3: BAR 13: no space for [io  size 0x1000]
[    0.824386] pci 0000:00:15.3: BAR 13: failed to assign [io  size 0x1000]
[    0.825393] pci 0000:00:15.4: BAR 13: no space for [io  size 0x1000]
[    0.826387] pci 0000:00:15.4: BAR 13: failed to assign [io  size 0x1000]
[    0.827390] pci 0000:00:15.5: BAR 13: no space for [io  size 0x1000]
[    0.828387] pci 0000:00:15.5: BAR 13: failed to assign [io  size 0x1000]
[    0.829391] pci 0000:00:15.6: BAR 13: no space for [io  size 0x1000]
[    0.830385] pci 0000:00:15.6: BAR 13: failed to assign [io  size 0x1000]
[    0.831409] pci 0000:00:15.7: BAR 13: no space for [io  size 0x1000]
[    0.832387] pci 0000:00:15.7: BAR 13: failed to assign [io  size 0x1000]
[    0.833394] pci 0000:00:16.3: BAR 13: no space for [io  size 0x1000]
[    0.834388] pci 0000:00:16.3: BAR 13: failed to assign [io  size 0x1000]
[    0.835402] pci 0000:00:16.4: BAR 13: no space for [io  size 0x1000]
[    0.836392] pci 0000:00:16.4: BAR 13: failed to assign [io  size 0x1000]
[    0.837378] pci 0000:00:16.5: BAR 13: no space for [io  size 0x1000]
[    0.838373] pci 0000:00:16.5: BAR 13: failed to assign [io  size 0x1000]
```

**Describe the results you expected:**
I would expect the container to unzip the file and echo "done unzipping"

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue consistently happens on v1.12.2 and v1.12.5

**Output of `docker version`:**

```
Client:
 Version:      1.12.5
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   7392c3b
 Built:        Fri Dec 16 02:23:59 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.5
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   7392c3b
 Built:        Fri Dec 16 02:23:59 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 7
 Running: 5
 Paused: 0
 Stopped: 2
Images: 6
Server Version: 1.12.5
Storage Driver: btrfs
 Build Version: Btrfs v4.4.1
 Library Version: 101
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.2 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 23.39 GiB
Name: lab-8
ID: OU7Y:HSGA:F5PN:QY3Z:QQTK:WFWT:FM6Y:WTSL:CQAL:QKH3:C2CY:KBH5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on a private cloud vmware esxi 6.0 with Rhel 7.2 guest.

**Comment by thaJeztah** [comment]
ping @tonistiigi @mlaventure any ideas?

**Comment by tonistiigi** [comment]
I don't even understand why `gzip -d` should be memory bound. It should use very little memory.

**Comment by ibrokethecloud** [comment]
Well this is just a sample. I had another container which unzips a file but i cannot share the docker file because it has organisation specific information.

Hence i had to come up with this dummy image. 

The same image works fine on 1.10.3

**Comment by ibrokethecloud** [comment]
I think the PCI errors might be a red herring.

I turned up debug mode for the docker daemon and re-ran the same container.

docker events generates the following output:
```2017-01-10T14:07:58.301017421+11:00 container create cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17 (image=test:1, name=adoring_raman)
2017-01-10T14:07:58.302443963+11:00 container attach cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17 (image=test:1, name=adoring_raman)
2017-01-10T14:07:58.336518729+11:00 network connect 5987cb637ae8b16cb803fe8eb8f5658a00003bf4552497e6875c8c3f23da6ed9 (container=cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17, name=bridge, type=bridge)
2017-01-10T14:07:58.564095395+11:00 container start cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17 (image=test:1, name=adoring_raman)
2017-01-10T14:07:58.566825792+11:00 container resize cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17 (height=49, image=test:1, name=adoring_raman, width=130)
```

While the docker daemon only mentions the following before the host reboots:
```
Jan 10 14:07:58 localhost NetworkManager[557]: <info>  (docker0): bridge port veth88e8c2c was attached
Jan 10 14:07:58 localhost NetworkManager[557]: <info>  (veth88e8c2c): enslaved to docker0
Jan 10 14:07:58 localhost dockerd: time="2017-01-10T14:07:58.318042744+11:00" level=debug msg="Assigning addresses for endpoint adoring_raman's interface on network bridge"
Jan 10 14:07:58 localhost dockerd: time="2017-01-10T14:07:58.333414524+11:00" level=debug msg="Programming external connectivity on endpoint adoring_raman (5a0c208454a74a71b44c3420a1a3434c7790914f2adf1aa8ef024f491a4a2692)"
Jan 10 14:07:58 localhost kernel: docker0: port 1(veth88e8c2c) entered forwarding state
Jan 10 14:07:58 localhost kernel: docker0: port 1(veth88e8c2c) entered forwarding state
Jan 10 14:07:58 localhost NetworkManager[557]: <info>  (docker0): link connected
Jan 10 14:07:58 localhost dockerd: time="2017-01-10T14:07:58.460089613+11:00" level=debug msg="sandbox set key processing took 36.544841ms for container cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17"
Jan 10 14:07:58 localhost dockerd: time="2017-01-10T14:07:58.561211280+11:00" level=debug msg="libcontainerd: received containerd event: &types.Event{Type:\"start-container\", Id:\"cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17\", Status:0x0, Pid:\"\", Timestamp:(*timestamp.Timestamp)(0xc820c38d90)}"
Jan 10 14:07:58 localhost dockerd: time="2017-01-10T14:07:58.564334306+11:00" level=debug msg="libcontainerd: event unhandled: type:\"start-container\" id:\"cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17\" timestamp:<seconds:1484017678 nanos:560529137 > "
Jan 10 14:07:58 localhost dockerd: time="2017-01-10T14:07:58.565912665+11:00" level=debug msg="Calling POST /v1.24/containers/cd0c523f54e7f64e63b7af0f2ab7cd5ebf6a167861af51365683524ad9137f17/resize?h=49&w=130"
```

**Comment by mlaventure** [comment]
Could you share the output `lspci`?

I'm assuming that the BAR that fails to be assigned belong to the storage, just want to confirm.

Also, BARs are assigned during PCI enumeration, so when the machine boots (hence why the timestamp is less than 1sec). So if those BAR don't belong to a PCI device connected to the storage, they probably have nothing to do with the machine being rebooted

**Comment by ibrokethecloud** [comment]
Here we go.

```
00:00.0 Host bridge: Intel Corporation 440BX/ZX/DX - 82443BX/ZX/DX Host bridge (rev 01)
00:01.0 PCI bridge: Intel Corporation 440BX/ZX/DX - 82443BX/ZX/DX AGP bridge (rev 01)
00:07.0 ISA bridge: Intel Corporation 82371AB/EB/MB PIIX4 ISA (rev 08)
00:07.1 IDE interface: Intel Corporation 82371AB/EB/MB PIIX4 IDE (rev 01)
00:07.3 Bridge: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 08)
00:07.7 System peripheral: VMware Virtual Machine Communication Interface (rev 10)
00:0f.0 VGA compatible controller: VMware SVGA II Adapter
00:10.0 SCSI storage controller: LSI Logic / Symbios Logic 53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI (rev 01)
00:11.0 PCI bridge: VMware PCI bridge (rev 02)
00:15.0 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.1 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.2 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.3 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.4 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.5 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.6 PCI bridge: VMware PCI Express Root Port (rev 01)
00:15.7 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.0 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.1 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.2 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.3 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.4 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.5 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.6 PCI bridge: VMware PCI Express Root Port (rev 01)
00:16.7 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.0 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.1 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.2 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.3 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.4 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.5 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.6 PCI bridge: VMware PCI Express Root Port (rev 01)
00:17.7 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.0 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.1 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.2 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.3 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.4 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.5 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.6 PCI bridge: VMware PCI Express Root Port (rev 01)
00:18.7 PCI bridge: VMware PCI Express Root Port (rev 01)
02:00.0 USB controller: VMware USB1.1 UHCI Controller
02:01.0 SCSI storage controller: LSI Logic / Symbios Logic 53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI (rev 01)
02:02.0 USB controller: VMware USB2 EHCI Controller
02:04.0 SCSI storage controller: LSI Logic / Symbios Logic 53c1030 PCI-X Fusion-MPT Dual Ultra320 SCSI (rev 01)
03:00.0 Ethernet controller: VMware VMXNET3 Ethernet Controller (rev 01)
```

**Comment by mlaventure** [comment]
@ibrokethecloud thanks.

Those errors don't seem to be related to your storage devices.

Hence, I'd say that somehow the change in the way docker operates is triggering a bug in your kernel. Since there doesn't seem to be any logs saved, your kernel probably paniced.

You may want to try and update your kernel and see if that fixes it.

**Comment by ibrokethecloud** [comment]
So I have run the same OS image on AWS with docker 1.12.5 and I have hit the same problem.

**Comment by ibrokethecloud** [comment]
@thaJeztah @mlaventure i just downgraded to 1.12.0 and everything looks in order.

**Comment by ibuildthecloud** [comment]
While troubleshooting this issue I got the following.  I'm guessing this is just a kernel bug in RHEL 7.2 given that btrfs is "experimental".  I'll see if I can dig into this more.

```
[ 4134.776053] PGD 1fa0067 PUD 10fffd067 PMD 80000000e2a001e3 
[ 4134.776053] Thread overran stack, or stack corrupted
[ 4134.776053] Oops: 0011 [#1] SMP 
[ 4134.776053] Modules linked in: veth ipt_MASQUERADE nf_nat_masquerade_ipv4 iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 xt_addrtype iptable_filter xt_conntrack nf_nat nf_conntrack br_netfilter bridge stp llc isofs intel_powerclamp intel_rapl iosf_mbi crc32_pclmul ghash_clmulni_intel aesni_intel lrw gf128mul glue_helper ablk_helper cryptd ppdev pcspkr cirrus ttm drm_kms_helper parport_pc syscopyarea sysfillrect parport sysimgblt fb_sys_fops drm i2c_piix4 i2c_core ip_tables xfs libcrc32c btrfs zlib_deflate raid6_pq xor ata_generic pata_acpi crct10dif_pclmul crct10dif_common xen_netfront xen_blkfront crc32c_intel ata_piix serio_raw libata floppy fjes
[ 4134.776053] CPU: 0 PID: 2757 Comm: gzip Not tainted 3.10.0-514.2.2.el7.x86_64 #1
[ 4134.776053] Hardware name: Xen HVM domU, BIOS 4.2.amazon 12/09/2016
[ 4134.776053] task: ffff8800e5178000 ti: ffff8800e2b70000 task.ti: ffff8800e2b70000
[ 4134.776053] RIP: 0010:[<ffff8800e2b73fd8>]  [<ffff8800e2b73fd8>] 0xffff8800e2b73fd7
[ 4134.776053] RSP: 0018:ffff8800e2b6fb08  EFLAGS: 00010282
[ 4134.776053] RAX: 00000000fffffff2 RBX: ffff8800e2b73fd8 RCX: ffff8800e2b70000
[ 4134.776053] RDX: ffff8800e3b38000 RSI: 0000000000000020 RDI: ffff8800e2b73fd8
[ 4134.776053] RBP: ffff8800e2b73fd8 R08: ffff8800e2b70000 R09: 0000000000000000
[ 4134.776053] R10: 0000000000000000 R11: 0000000000000000 R12: ffff8800e2b73fd8
[ 4134.776053] R13: ffff8800e5178000 R14: 0000000000000086 R15: ffff8800e2b6fb70
[ 4134.776053] FS:  00007f4dc4dfd700(0000) GS:ffff88010b200000(0000) knlGS:0000000000000000
[ 4134.776053] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 4134.776053] CR2: ffff8800e2b73fd8 CR3: 00000000e2bdf000 CR4: 00000000001406f4
[ 4134.776053] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[ 4134.776053] DR3: 0000000000000000 DR6: 00000000ffff0ff0 DR7: 0000000000000400
[ 4134.776053] Stack:
[ 4134.776053]  ffffffff81696618 ffff8800e2b6fb70 0000000000000086 ffff8800e5178000
[ 4134.911869]  ffff8800e2b73fd8 ffff8800e2b73fd8 ffff8800e2b73fd8 ffff8800e5178000
[ 4134.911869]  ffff8800e2b70000 0000000000000002 ffff8800e2b6fd60 0000000000000020
[ 4134.911869] Call Trace:
[ 4134.911869] Code: ff ff ff ff ff ff ff ff ff ff ff 00 80 00 00 00 00 00 00 00 02 66 00 00 00 00 00 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 <80> b6 90 c4 4d 7f 00 00 33 00 00 00 00 00 00 00 46 02 00 00 00 
[ 4134.911869] RIP  [<ffff8800e2b73fd8>] 0xffff8800e2b73fd7
[ 4134.911869]  RSP <ffff8800e2b6fb08>
[ 4134.911869] CR2: ffff8800e2b73fd8
```

**Comment by ibrokethecloud** [comment]
Thanks a lot @ibuildthecloud . I have logged a support call with Redhat. Will update this issue as soon as I have some feedback.

**Comment by ibrokethecloud** [comment]
A new hotfix kernel has been released by redhat - 3.10.0-514.11.1.el7.x86_64 which contains the fix.

**Comment by thaJeztah** [comment]
Thanks for the update @ibrokethecloud !
